drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -2042,10 +2268,92 @@ bool RasterizerCanvasGLES2::try_join_item(Item *p_ci, RenderItemState &r_ris, bo
 		// it is possible, but not if they overlap, because
 		// a + light_blend + b + light_blend IS NOT THE SAME AS
 		// a + b + light_blend
-		join = false;
 
-		// we also dont want to allow joining this item with the next item, because the next item could have no lights!
-		r_batch_break = true;
+		bool light_allow_join = true;
+
+		// this is a quick getout if we have turned off light joining
+		if ((bdata.settings_light_max_join_items == 0) || r_ris.light_region.too_many_lights) {
+			light_allow_join = false;
+		} else {
+			// do light joining...
+
+			// first calculate the light bitfield
+			uint64_t light_bitfield = 0;
+			uint64_t shadow_bitfield = 0;
+			Light *light = r_ris.item_group_light;
+
+			int light_count = -1;
+			while (light) {
+				light_count++;
+				uint64_t light_bit = 1 << light_count;
+
+				// use the bounding rect of the joined items, NOT only the bounding rect of the first item.
+				// note this is a cost of batching, the light culling will be less effective
+
+				if (p_ci->light_mask & light->item_mask && r_ris.item_group_z >= light->z_min && r_ris.item_group_z <= light->z_max) {
+					//&& p_ci->global_rect_cache.intersects_transformed(light->xform_cache, light->rect_cache)) {
```

Comment:
Can remove old commented out code

Reply:
Ah yes, this is interesting actually. I left this in (especially during development) because I wasn't clear on whether the best choice for this may vary according to the situation, so I was continuously flipping between the 2 versions.

In my benchmark (which I must post here actually for you to try, it was essentially the sprite test I had posted earlier, but with a big 2d light over the central area), I found that with the random positions of sprites, the benefits of batching due to joining them irrespective of whether they were in the light area outweighed the cost of removing them from rendering if they were outside.

However there could be situations where doing this AABB test is better (I just haven't found any yet). This is something that may be revisited, I think it would be better to put a comment to this effect. I'm wary of making it a project setting at this stage, as it may not be necessary, and we already have a lot of settings.

Reply:
Ok I've cleared up the comments and made them a bit more sensible. :+1: